waspc/src/Wasp/Generator/Setup.hs

```diff
@@ -1,47 +1,21 @@
 module Wasp.Generator.Setup
-  ( setup,
+  ( runSetup,
   )
 where
 
-import Control.Concurrent (Chan, newChan, readChan)
-import Control.Concurrent.Async (concurrently)
 import StrongPath (Abs, Dir, Path')
-import System.Exit (ExitCode (..))
+import Wasp.AppSpec (AppSpec)
 import Wasp.Generator.Common (ProjectRootDir)
-import qualified Wasp.Generator.Job as J
-import Wasp.Generator.Job.IO (printPrefixedJobMessage)
-import Wasp.Generator.ServerGenerator.Setup (setupServer)
-import Wasp.Generator.WebAppGenerator.Setup (setupWebApp)
+import qualified Wasp.Generator.DbGenerator as DbGenerator
+import Wasp.Generator.Monad (GeneratorError (..), GeneratorWarning (..))
+import Wasp.Generator.NpmInstall (ensureNpmInstall)
 
-setup :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
-setup projectDir = do
-  chan <- newChan
-  let runSetupJobs = concurrently (setupServer projectDir chan) (setupWebApp projectDir chan)
-  (_, result) <- concurrently (handleJobMessages chan) runSetupJobs
-  case result of
-    (ExitSuccess, ExitSuccess) -> return $ Right ()
-    exitCodes -> return $ Left $ setupFailedMessage exitCodes
-  where
-    handleJobMessages = go Nothing (False, False)
-      where
-        go :: Maybe J.JobMessage -> (Bool, Bool) -> Chan J.JobMessage -> IO ()
-        go _ (True, True) _ = return ()
-        go prevJobMsg (isWebAppDone, isServerDone) chan = do
-          jobMsg <- readChan chan
-          case J._data jobMsg of
-            J.JobOutput {} ->
-              printPrefixedJobMessage prevJobMsg jobMsg
-                >> go (Just jobMsg) (isWebAppDone, isServerDone) chan
-            J.JobExit {} -> case J._jobType jobMsg of
-              J.WebApp -> go (Just jobMsg) (True, isServerDone) chan
-              J.Server -> go (Just jobMsg) (isWebAppDone, True) chan
-              J.Db -> error "This should never happen. No db job should be active."
-
-    setupFailedMessage (serverExitCode, webAppExitCode) =
-      let serverErrorMessage = case serverExitCode of
-            ExitFailure code -> " Server setup failed with exit code " ++ show code ++ "."
-            _ -> ""
-          webAppErrorMessage = case webAppExitCode of
-            ExitFailure code -> " Web app setup failed with exit code " ++ show code ++ "."
-            _ -> ""
-       in "Setup failed!" ++ serverErrorMessage ++ webAppErrorMessage
+runSetup :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+runSetup spec dstDir = do
+  -- TODO: Use a monad transformer here in the future for better shortcutting behavior
```

Comment:
Ok, although this is so simple all together that I wouldn't worry about it too much. But I guess the idea was to re-use Generator monad in some way since it already does this exact kind of logic -> end on error, accumulate warnings?